### PR TITLE
[Bug] [PIR] [NameHandling] Use marking in compilation

### PR DESCRIPTION
**Pull request by effectfully**
Don't look here yet.

### CODE REVIEW

**Comment by effectfully** [comment]
So I was curious if #3416 fixed that problem reported by @jakzale where adding `markNonFreshTerm` to the PIR renamer breaks the tests and it didn't. Hence I started digging. Fortunately, adding `markNonFrestTerm` breaks a rather simple test and so I was able to check manually that the renamed term is equal to the original one. My next guess was to check if adding `markNonFrestTerm` to the compilation pipeline as well could help and it indeed did. So previously the compiler would generate "fresh" names clashing with already existing ones and now at least one instance of this problem is eliminated.

@jakzale check if that fixes your bug as well.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs [hunk 1]
```diff
@@ -58,9 +59,8 @@ Two problems arise:
 -}
 
 instance PLC.HasUniques (Term tyname name uni fun ann) => PLC.Rename (Term tyname name uni fun ann) where
-    -- TODO: the Plutus Core codebase uses marking in order to prevent clashing with existing
-    -- free variables. Should we do the same here?
-    rename = PLC.runRenameT . renameTermM
+    -- See Note [Marking]
+    rename = through markNonFreshTerm >=> PLC.runRenameT . renameTermM
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs:63]: Hmm, I suggested that Jakub try this in the past and it made things fail. I guess it doesn't here.

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs [hunk 2]
```diff
@@ -90,7 +91,10 @@ compileToReadable =
 -- Note: the result *does* have globally unique names.
 compileReadableToPlc :: Compiling m e uni fun a => Term TyName Name uni fun (Provenance a) -> m (PLCTerm uni fun a)
 compileReadableToPlc =
-    NonStrict.compileNonStrictBindings
+    -- See Note [Marking]. Failing to call 'markNonFreshTerm' results in variable capture
+    -- even in simple cases causing tests to fail.
+    through markNonFreshTerm
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:96]: I don't see why putting this here changes anything. This gets run after `compileToReadable`, which starts with `rename`, which given your other change should get marking right. It's weird that we could somehow get here and have problems...

Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:96]: Oh, indeed. I have a hypothesis though: both `inline` and `removeDeadBindings` call `runQuoteT` and so we lose the information of what was the unused unique and start from the same place where we were before calling `inline` and `removeDeadBindings`, which is how we get variable capture. I'll try dropping `runQuoteT` from both of these functions.

Comment (michaelpj) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:96]: Yeah, that sounds right. They probably shouldn't do that.

File: plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/Rename.hs:63]: Yes, see my comment above.

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs]: I've dropped those `runQuoteT`s, rereview.

**Review by michaelpj — APPROVED** [review]

**Review by michaelpj — APPROVED** [review]

**Pull request merged**